pkg/yurttunnel/pki/certmanager/certmanager.go

```diff
@@ -17,21 +17,345 @@ limitations under the License.
+		if err != nil {
+			return dnsNames, ips, err
+		}
+		ips = append(ips, net.ParseIP(host))
+	}
+	return dnsNames, ips, nil
+}
+
+// getClusterIPDNSandIP gets the DNS names and IPs from the NodePort service
+func getNodePortDNSandIP(
+	clientset kubernetes.Interface,
+	dnsNames []string,
+	ips []net.IP) ([]string, []net.IP, error) {
+	labelSelector := fmt.Sprintf("%s=false", constants.YurtEdgeNodeLabel)
+	// yurttunnel-server will be deployed on one of the cloud nodes
+	nodeLst, err := clientset.CoreV1().Nodes().List(
+		metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return dnsNames, ips, err
+	}
+	if len(nodeLst.Items) == 0 {
+		return dnsNames, ips, errors.New("there is no cloud node")
+	}
+	for i, addr := range nodeLst.Items[0].Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			ips = append(ips, net.ParseIP(addr.Address))
+		}
+		// there is no qualified address (i.e. NodeInternalIP)
+		if i == len(nodeLst.Items[0].Status.Addresses)-1 {
+			return dnsNames, ips, errors.New("can't find node IP")
+		}
+	}
+	return dnsNames, ips, nil
+}
+
+// NewYurttunnelAgentCertManager creates a certificate manager for
+// the yurttunel-agent
+func NewYurttunnelAgentCertManager(
+	clientset kubernetes.Interface) (certificate.Manager, error) {
+	nodeIP, err := getAgentNodeIP(clientset, os.Getenv("NODE_NAME"))
+	if err != nil {
+		return nil, err
+	}
+
+	return newCertManager(
+		clientset,
+		"yurttunnel-agent",
+		constants.YurttunnelAgentCertDir,
+		constants.YurttunnelAgentCSRCN,
+		[]string{constants.YurttunnelCSROrg},
+		[]string{os.Getenv("NODE_NAME")},
+		[]net.IP{net.ParseIP(nodeIP)})
+}
+
+// getAgentNodeIP returns the IP of the node whose name is nodeName
+func getAgentNodeIP(
+	clientset kubernetes.Interface,
+	nodeName string) (string, error) {
+	node, err := clientset.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("fail to get node(%s): %s", nodeName, err)
+		return "", err
+	}
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			return addr.Address, nil
+		}
+	}
+	klog.Errorf("node(%s) doesn't have internalIP", nodeName)
+	return "", fmt.Errorf("internalIP of node(%s) is not found", nodeName)
+}
+
 // NewCertManager creates a certificate manager that will generates a
 // certificate by sending a csr to the apiserver
-func NewCertManager(
+func newCertManager(
+	clientset kubernetes.Interface,
+	componentName,
+	certDir,
+	commonName string,
+	organizations,
+	dnsNames []string, ipAddrs []net.IP) (certificate.Manager, error) {
+	certificateStore, err :=
+		certificate.NewFileStore(componentName, certDir, certDir, "", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the server certificate store: %v", err)
+	}
+
+	getTemplate := func() *x509.CertificateRequest {
+		return &x509.CertificateRequest{
+			Subject: pkix.Name{
+				CommonName:   constants.YurttunneServerCSRCN,
```

Comment:
constants.YurttunneServerCSRCN should be commonName?

Reply:
Yes. it is the common name of the server's certificate.

Reply:
fixed